registry/client/repository.go

```diff
@@ -156,74 +156,140 @@ func (r *repository) Manifests(ctx context.Context, options ...distribution.Mani
 	}, nil
 }
 
-func (r *repository) Signatures() distribution.SignatureService {
-	ms, _ := r.Manifests(r.context)
-	return &signatures{
-		manifests: ms,
+func (r *repository) Tags(ctx context.Context) distribution.TagService {
+	return &tags{
+		client:  r.client,
+		ub:      r.ub,
+		context: r.context,
+		name:    r.Name(),
 	}
 }
 
-type signatures struct {
-	manifests distribution.ManifestService
-}
-
-func (s *signatures) Get(dgst digest.Digest) ([][]byte, error) {
-	m, err := s.manifests.Get(dgst)
-	if err != nil {
-		return nil, err
-	}
-	return m.Signatures()
+// tags implements remote tagging operations.
+type tags struct {
+	client  *http.Client
+	ub      *v2.URLBuilder
+	context context.Context
+	name    string
 }
 
-func (s *signatures) Put(dgst digest.Digest, signatures ...[]byte) error {
-	panic("not implemented")
-}
+// All returns all tags
+func (t *tags) All(ctx context.Context) ([]string, error) {
+	var tags []string
 
-type manifests struct {
-	name   string
-	ub     *v2.URLBuilder
-	client *http.Client
-	etags  map[string]string
-}
-
-func (ms *manifests) Tags() ([]string, error) {
-	u, err := ms.ub.BuildTagsURL(ms.name)
+	u, err := t.ub.BuildTagsURL(t.name)
 	if err != nil {
-		return nil, err
+		return tags, err
 	}
 
-	resp, err := ms.client.Get(u)
+	resp, err := t.client.Get(u)
 	if err != nil {
-		return nil, err
+		return tags, err
 	}
 	defer resp.Body.Close()
 
 	if SuccessStatus(resp.StatusCode) {
 		b, err := ioutil.ReadAll(resp.Body)
 		if err != nil {
-			return nil, err
+			return tags, err
 		}
 
 		tagsResponse := struct {
 			Tags []string `json:"tags"`
 		}{}
 		if err := json.Unmarshal(b, &tagsResponse); err != nil {
-			return nil, err
+			return tags, err
 		}
+		tags = tagsResponse.Tags
+		return tags, nil
+	}
+	return tags, handleErrorResponse(resp)
+}
+
+func descriptorFromHeaders(headers http.Header) (distribution.Descriptor, error) {
+	desc := distribution.Descriptor{}
 
-		return tagsResponse.Tags, nil
+	digestHeader := headers.Get("Docker-Content-Digest")
```

Comment:
Adding back commentary that was hidden as an outdated diff
stevvooe   Collaborator
This header is not required.

RichardScothern   Collaborator
RichardScothern added a note 21 hours ago
This function needs to read this header to make the digest portion of the descriptor

stevvooe   Collaborator
stevvooe added a note 20 hours ago
Okay, so users of this function need to fall back to another source if this can't be resolved by header. This is not guaranteed to be set.

RichardScothern   Collaborator
RichardScothern added a note 18 hours ago
Hmmm, the spec indicates the Docker-Content-Digest header would be set (and the code includes it as far back as v2.0.0) @stevvooe